File: garnet/drivers/bluetooth/lib/sm/packet.cc
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "packet.h"

namespace btlib {
namespace sm {

PacketReader::PacketReader(const common::ByteBuffer* buffer)
    : common::PacketView<Header>(buffer, buffer->size() - sizeof(Header)) {}

PacketWriter::PacketWriter(Code code, common::MutableByteBuffer* buffer)
    : common::MutablePacketView<Header>(buffer,
                                        buffer->size() - sizeof(Header)) {
  mutable_header()->code = code;
}

}  // namespace sm
}  // namespace btlib

File: src/connectivity/network/netstack3/src/bindings/socket/packet.rs
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use fidl_fuchsia_posix as fposix;
use fidl_fuchsia_posix_socket_packet as fppacket;
use futures::{future, TryStreamExt as _};

pub(crate) async fn serve(stream: fppacket::ProviderRequestStream) -> Result<(), fidl::Error> {
    stream
        .try_for_each(|req| {
            match req {
                fppacket::ProviderRequest::Socket { responder, kind: _ } => {
                    log::error!("TODO(https://fxbug.dev/106735): Support packet sockets");
                    responder_send!(responder, &mut Err(fposix::Errno::Enoprotoopt));
                }
            };
            future::ok(())
        })
        .await
}
